### PR TITLE
Enable multiple files to be included in prebuild packages

### PR DESCRIPTION
**Pull request by daviwil**
This PR finishes the work started by @tmpvar in #122 and @antoinepairet in #171.  It accomplishes two goals:

- Allowing multiple `.node` files to be included in a prebuilt binary package
- Allowing the user to specify a regex using `--collect-files-filter` to be used for including other files in the build output path

A corresponding PR for prebuild-install will be sent later.

### CODE REVIEW

**Comment by daviwil** [comment]
There's a test failure I need to look into, plus looking into @todbot's issue.

**Comment by daviwil** [comment]
@todbot, can you try again with my branch to see if you can still reproduce your issue?  I just tried it on my Windows machine and couldn't reproduce it.  I'll try it on my MacBook in a moment to see if it shows up there.  I used this command in the `node-hid` repo that I cloned:

```
prebuild -t 7.0.0 --collect-files-filter "\.(node|lib)$"
```

The resulting .tar.gz file had the expected files, `HID.lib`, `HID.node`, and `hidapi.lib`.

@ralphtheninja I think this PR is ready for review!  Can you check the documentation I added to README.md to see if it's sufficient?  Also, do you want to settle on the `--collect-files-filter` parameter name or possibly change to something like `--included-files-filter`?

File: help.txt [hunk 1]
```diff
@@ -13,3 +13,4 @@ prebuild [options]
   --debug                       (set Debug or Release configuration)
   --verbose                     (log verbosely)
   --version                     (print prebuild version and exit)
+  --collect-files-filter        (regex to match files that will be distributed [default: '\\.(node|a)$'])
```
Comment (ralphtheninja) [inline help.txt]: `[default: '\\.(node|a)$']` should be `[default: '\\.(node)$']` according to the value in `rc.js`

**Comment by ralphtheninja** [comment]
> @ralphtheninja I think this PR is ready for review! Can you check the documentation I added to README.md to see if it's sufficient? Also, do you want to settle on the --collect-files-filter parameter name or possibly change to something like --included-files-filter?

Naming is always the most difficult thing. I'm thinking maybe we could shorten it a bit? What about just `--include` or something else that is short and to the point.

I'd like to get some input from other maintainers as well.

Other than that, this looks great!

File: README.md [hunk 2]
```diff
@@ -41,6 +41,21 @@ See [`allTargets`](https://github.com/lgeiger/node-abi#usage) for all available
 
 For more options run `prebuild --help`. The prebuilds created are compatible with [node-pre-gyp](https://github.com/mapbox/node-pre-gyp)
 
+If you'd like to include other files with your prebuilds like additional
+`.node` files or other native libraries, you can pass a file-matching regular
+expression to the  `--collect-files-filter` parameter:
```
Comment (ralphtheninja) [inline README.md]: you can pass a file-matching regular expression to the  `--collect-files-filter` parameter

->

you can pass a file-matching regular expression to `--collect-files-filter`:

File: README.md [hunk 3]
```diff
@@ -41,6 +41,21 @@ See [`allTargets`](https://github.com/lgeiger/node-abi#usage) for all available
...
+
+```
+prebuild -t 8.0.0 --collect-files-filter "\.(node|a)$"
+```
+
+Note that if you include multiple `.node` files, you will need to use the
+prebuild-install's '`--binary-name` parameter to indicate which file should be
```
Comment (ralphtheninja) [inline README.md]: Looks like there's a `'` hanging loose there, just before `--binary-name` :)

File: help.txt [hunk 4]
```diff
@@ -7,10 +7,10 @@ prebuild [options]
   --upload-all  -u  [gh-token]  (upload all files from ./prebuilds folder to github)
   --preinstall  -i  script      (run this script before prebuilding)
   --path        -p  path        (make a prebuild here)
+  --collect-files-filter        (regex to match files that will be distributed [default: '\.node$'])
```
Comment (ralphtheninja) [inline help.txt]: Good that you moved this up a bit!

File: test/collect-artifacts-test.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    t.error(err, 'collected files')
+    t.equal(collected.length, 2, 'found two files')
+
+    // ensure the files are always in the same order
+    collected.sort()
+
+    t.deepEqual(collected, [
+      path.join(release, 'test.node'),
+      path.join(release, 'test.out')
+    ])
```
Comment (ralphtheninja) [inline test/collect-artifacts-test.js]: :+1:

File: help.txt (hunk 1 above)
Comment (daviwil) [reply help.txt]: Yep, already caught that!

File: README.md (hunk 2 above)
Comment (daviwil) [reply README.md]: 👍

File: README.md (hunk 3 above)
Comment (daviwil) [reply README.md]: Whoops, thanks!  Also, what do you think about `--binary-name` as the `prebuild-install` parameter for choosing the right .node file?  Is there a more obvious name?

**Comment by daviwil** [comment]
I definitely like `--include`, but should it be `--include-regex` just to be super clear about what the parameter accepts?  If you think it's not necessary, I'm happy to go with `--include`.

**Comment by ralphtheninja** [comment]
> I definitely like --include, but should it be --include-regex just to be super clear about what the parameter accepts? If you think it's not necessary, I'm happy to go with --include.

I think `--include-regex` is fine as well.

**Comment by lgeiger** [comment]
This looks great! Thanks everybody for the great work! :tada:

I think using either `--include-regex` or `--include-pattern` would be a good choice.

**Comment by ralphtheninja** [comment]
@juliangruber Do you have any input/comments?

**Comment by juliangruber** [comment]
This looks good to me! I didn't sing off on it because you had requested changes still

**Comment by daviwil** [comment]
@lgeiger `--include-pattern` is a great suggestion, but maybe we should save that in case we also want to support `minimatch` [globbing patterns](https://www.npmjs.com/package/minimatch#usage) in the future?

Just pushed an update with the requested fixes. @ralphtheninja, can you check once again?   Before merging this I'd also like to get confirmation on the `prebuild-install --binary-name` parameter that I invented for the new documentation.  Based on what you decide, I'll send a PR to `prebuild-install` later today to add that parameter.

**Comment by ralphtheninja** [comment]
@juliangruber Approved.

@daviwil `--binary-name` is fine imo.

> I'll send a PR to prebuild-install later today to add that parameter.

After this is merged we can hold off with release until `prebuild-install` is also on par, so we can release them in sync.

**Comment by ralphtheninja** [comment]
Screw holding off the release. This is backwards compatible. Releasing it as `7.1.0`.

**Comment by daviwil** [comment]
Awesome, thanks a lot! The prebuild-install PR will be out as soon as I write a test :)

**Comment by ralphtheninja** [comment]
> Awesome, thanks a lot! The prebuild-install PR will be out as soon as I write a test :)

Thank you! :grinning:

**Comment by lgeiger** [comment]
Excellent thanks for getting this shipped.

**Pull request merged**